File: go/quicksortInt.go
func quicksortInt(array []int, start int, end int) {
	if start >= end {
		return
	}
	pivot := array[end]
	left := 0
	right := 0
	for left+right < end-start {
		num := array[start+left]
		if num < pivot {
			left++
		} else {
			array[start+left] = array[end-right-1]
			array[end-right-1] = pivot
			array[end-right] = num
			right++
		}
	}
	quicksortInt(array, start, start+left-1)
	quicksortInt(array, start+left+1, end)
}

File: rust/src/quicksortInt.rs
fn quicksort_int(array: &mut [i32], start: usize, end: usize) {
    if start >= end {
        return;
    }
    let pivot = array[end];
    let mut left = 0;
    let mut right = 0;
    while left + right < end - start {
        let num = array[start + left];
        if num < pivot {
            left += 1;
        } else {
            array.swap(start + left, end - right - 1);
            array.swap(end - right - 1, end - right);
            right += 1;
        }
    }
    quicksort_int(array, start, start + left - 1);
    quicksort_int(array, start + left + 1, end);
}
